third_party/amd/lib/TritonAMDGPUToLLVM/DotOpToLLVM/MFMA.cpp

```diff
@@ -387,109 +371,120 @@ struct DotOpMFMAConversionHelper {
-          // This is only for the operands of scaled mfma on CDNA4
-          results.push_back(b.bitcast(vec, vec_ty(i32_ty, 4)));
-        if (32 == kBase)
-          results.push_back(b.bitcast(vec, vec_ty(i32_ty, 8)));
-      } else {
-        results.push_back(vec);
       }
+      if (4 == kBase)
+        // This is for int8 on pre- CDNA3 GPUs
+        results = b.bitcast(vec, i32_ty);
+      if (8 == kBase)
+        results = b.bitcast(vec, i64_ty);
+      if (16 == kBase)
+        // This is only for the operands of scaled mfma on CDNA4
+        results = b.bitcast(vec, vec_ty(i32_ty, 4));
+      if (32 == kBase)
+        results = b.bitcast(vec, vec_ty(i32_ty, 8));
+    } else {
+      results = vec;
     }
     return results;
   }
 
   /// Converts dot operand structure to value table and converts types
   /// appropriate for mfma instructions
-  virtual SmallVector<ValueTable> getValuesFromDotOperandLayoutStruct(
-      Value value, int batch, int n0, int n1, int kWidth, int kBase, Type type,
-      bool allowXF32, bool preserveBF16, bool isConstantScale = false) const {
+  virtual ValueTable getValuesFromDotOperandLayoutStruct(
+      Value value, int batch, int nonKRep, int kRepInKWidth, int kWidth,
+      int kBase, Type type, bool allowXF32, bool preserveBF16,
+      bool isConstantScale = false) const {
     auto tb = TritonLLVMOpBuilder(loc, rewriter);
     auto elems = unpackLLElements(loc, value, rewriter);
-    int kpack = kWidth / kBase;
-    SmallVector<ValueTable> dotOpVals(kpack);
+    // number of kBase-element vectors
+    int numVecInKBase = kRepInKWidth * kWidth / kBase;
+    ValueTable dotOpVals;
     for (int b = 0; b < batch; ++b) {
-      for (int i = 0; i < n0; i++) {
-        for (int j = 0; j < n1; j++) {
+      for (int nonK = 0; nonK < nonKRep; nonK++) {
+        for (int kBaseVec = 0; kBaseVec < numVecInKBase; kBaseVec++) {
+          // For each kBase-element vector
+
+          // Step 1: construct each kBase-element vector by
+          //         - extracting kBase elements from elems and
+          //         - putting them into a kBase-element vector, i.e. rawElems
           Type elemTy = typeConverter->convertType(type);
-          Type ty = vec_ty(elemTy, kWidth);
+          Type ty = vec_ty(elemTy, kBase);
           Value rawElems = tb.undef(ty);
-          for (int k = 0; k < kWidth; ++k) {
-            rawElems = tb.insert_element(
-                ty, rawElems,
-                elems[kWidth * n1 * n0 * b + kWidth * n1 * i + kWidth * j + k],
-                tb.i32_val(k));
-          }
-
-          Value convertedElems;
+          for (int k = 0; k < kBase; ++k)
+            rawElems =
+                tb.insert_element(ty, rawElems,
+                                  elems[(b * nonKRep * numVecInKBase +
+                                         nonK * numVecInKBase + kBaseVec) *
+                                            kBase +
+                                        k],
```

Comment:
Nit: what about using a local variable to capture the result of `(b * nonKRep * numVecInKBase + nonK * numVecInKBase + kBaseVec) * kBase + k` first? Right now it spans 4 lines which does not look nicely..

To be fancy, this is `mlir::linearize({b, nonK, kBaseVec, k}, mlir::computeSuffixProduct({batch, nonKRep, numVecInKBase, kBase})` I think. But not sure it's more clear than this; so more wanted to point out some utility functions in [IndexingUtils.h](https://mlir.llvm.org/doxygen/IndexingUtils_8h_source.html). :)